core/src/test/java/io/neow3j/protocol/core/OfflineServiceTest.java

```diff
@@ -0,0 +1,111 @@
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class OfflineServiceTest {
+
+    private final String EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE = "Invalid neow3j service for this function. You " +
+            "cannot interact with a Neo node using an OfflineService instance.";
+
+    private final Neow3j neow3j = Neow3j.build();
+    private final Neow3jService service = OfflineService.newInstance();
+
+    @Test
+    public void testSend() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class,
+                () -> service.send(new Request(), NeoGetBlock.class));
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    @Test
+    public void testSendAsync() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class,
+                () -> service.sendAsync(new Request(), NeoGetBlock.class));
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    @Test
+    public void testSubscribeWithOnError() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class,
+                () -> service.subscribe(new Request(), "", Notification.class));
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    @Test
+    public void testClose() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class, service::close);
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    // Tests with a Neow3j instance that has an OfflineService
+
+    @Test
+    public void testNeow3jSend() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class,
+                () -> neow3j.getBestBlockHash().send());
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    @Test
+    public void testNeow3jSendAsync() {
+        OfflineServiceException thrown = assertThrows(OfflineServiceException.class,
+                () -> neow3j.getBestBlockHash().sendAsync());
+        assertThat(thrown.getMessage(), containsString(EXPECTED_OFFLINESERVICE_EXCEPTION_MESSAGE));
+    }
+
+    @Test
+    public void testNeow3jSubscribeWithOnError() throws IOException, InterruptedException {
+        AtomicBoolean onErrorExecuted = new AtomicBoolean(false);
+        neow3j.subscribeToNewBlocksObservable(true).subscribe(NeoGetBlock::getBlock, (e) -> onErrorExecuted.set(true));
+        Thread.sleep(100);
```

Comment:
`Thread.sleep()` in tests is a [code smell](https://en.wikipedia.org/wiki/Code_smell). I know we have places where we use this.
However, let's try to not introduce more and more on our codebase.
Can we/you think of a strategy to remove this `Thread.sleep()` in all the tests here?

Reply:
I'm aware it's a code smell. I haven't found a suitable way so far to wait on this disposable thread to dispose, thus, I just used sleep. I can dig a bit more.

Reply:
As communicated internally, these tests with observables are not needed in the scope of this test as they call `Neow3jService.send()` underlying which is already covered in this test class.